espresso/hotshot-listener/hotshot_listener.go

```diff
@@ -0,0 +1,267 @@
+	rollupSequencerManager            *espressogen.IEspressoRollupSequencerManager
+	quorumViewNumberBuilderCommitment map[string]big.Int
+	daViewNumberBuilderCommitment     map[string]bool
+	sequencerAddress                  string
+	conn                              *websocket.Conn
+}
+
+func NewHotshotListener(hotshotUrl string, rollupSequencerManagerContract string, l1Client *ethclient.Client, sequencerAddress string) (*HotshotListener, error) {
+
+	if hotshotUrl == "" {
+		return nil, fmt.Errorf("hotshot url is empty, please provide a valid url")
+	}
+	if rollupSequencerManagerContract == "" {
+		return nil, fmt.Errorf("rollup sequencer manager contract address is empty, please provide a valid address")
+	}
+	if sequencerAddress == "" {
+		return nil, fmt.Errorf("sequencer address is empty, please provide a valid address")
+	}
+	if l1Client == nil {
+		return nil, fmt.Errorf("l1 client is nil, please provide a valid client")
+	}
+
+	// Convert rollupSequencerManagerContract to an address
+	rollupSequencerManagerContractAddress := common.HexToAddress(rollupSequencerManagerContract)
+	rollupSequencerManager, err := espressogen.NewIEspressoRollupSequencerManager(rollupSequencerManagerContractAddress, l1Client)
+	if err != nil {
+		log.Error("failed to create rollup sequencer manager contract instance", "err", err)
+		return nil, err
+	}
+
+	// Create a new rollup sequencer manager contract instance
+	return &HotshotListener{
+		hotshotUrl:                        hotshotUrl + HotshotListenerEndpoint,
+		rollupSequencerManager:            rollupSequencerManager,
+		quorumViewNumberBuilderCommitment: make(map[string]big.Int),
+		daViewNumberBuilderCommitment:     make(map[string]bool),
+		sequencerAddress:                  sequencerAddress,
+	}, nil
+}
+
+func (listener *HotshotListener) processMessage(message []byte) error {
+	// Convert message to ConsensusMessage
+	consensusMessage, err := types.UnmarshalConsensusMessage(message)
+	if err != nil {
+		log.Error("failed to unmarshal consensus message:", err)
+		return err
+	}
+	// Quorum proposal represents a proposal that needs to be supported by a quorum of nodes
+	// this quorum proposal needs to be for a given view and builder commitment
+	if consensusMessage.Event.QuorumProposalWrapper != nil {
+		return listener.processQuorumProposalEvent(consensusMessage.Event.QuorumProposalWrapper)
+	}
+	// DA proposal event indicates thats data availability information
+	// is available for a given block with the given view number and builder commitment
+	if consensusMessage.Event.DaProposalWrapper != nil {
+		return listener.processDaProposalEvent(consensusMessage.Event.DaProposalWrapper)
+	}
+
+	// Only when hotshot builder has both quorum proposal and DA proposal for a given view
+	// it begins constructing another block
+	// Decide event in hotshot is the event when
+	// a view has been finalized by hotshot and cannot change now
+	if consensusMessage.Event.Decide != nil {
```

Comment:
Why does this event have no suffix `Wrapper` like others?

Reply:
not sure! This is how they have it in espresso-network